protocols/imap/src/main/java/org/apache/james/imap/decode/parser/GetAnnotationCommandParser.java

```diff
@@ -29,119 +28,99 @@
+                builder.maxsize(consumeMaxsizeOpt(requestReader));
                 break;
 
             case 'D':
-                getAnnotationRequest.setDepthValue(consumeDepthOpt(request));
+                builder.depth(consumeDepthOpt(requestReader));
                 break;
 
             default:
-                getAnnotationRequest.setKeys(consumeKeys(request));
+                builder.keys(consumeKeys(requestReader));
+                shouldEndOfLine = true;
                 break;
         }
-        request.consumeChar(')');
-        if (!getAnnotationRequest.getKeys().isEmpty()) {
-            request.eol();
+        requestReader.consumeChar(')');
+        if (shouldEndOfLine) {
+            requestReader.eol();
         }
     }
 
-    private Depth consumeDepthOpt(ImapRequestLineReader request) throws DecodingException {
-        request.consumeWord(new CharacterValidator() {
-            int pos = 0;
-            public boolean isValid(char chr) {
-                if (pos >= DEPTH.length) {
-                    return false;
-                } else {
-                    return DEPTH[pos++] == ImapRequestLineReader.cap(chr);
-                }
-            }
-        });
-        return Depth.fromString(request.atom());
-    }
-    public static void main(String[] agrs) {
-        System.out.println(Depth.fromString("-1"));
+    private Depth consumeDepthOpt(ImapRequestLineReader requestReader) throws DecodingException {
+        if (requestReader.atom().equalsIgnoreCase(DEPTH)) {
+            return Depth.fromString(requestReader.atom());
+        }
+
+        throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Wrong on, it should be DEPTH");
     }
 
-    private long consumeMaxsizeOpt(ImapRequestLineReader request) throws DecodingException {
-        request.consumeWord(new CharacterValidator() {
-            int pos = 0;
-            public boolean isValid(char chr) {
-                if (pos >= MAXSIZE.length) {
-                    return false;
-                } else {
-                    return MAXSIZE[pos++] == ImapRequestLineReader.cap(chr);
-                }
-            }
-        });
-        return request.number(true);
+    private long consumeMaxsizeOpt(ImapRequestLineReader requestReader) throws DecodingException {
+        if (requestReader.atom().equalsIgnoreCase(MAXSIZE)) {
+            return requestReader.number(true);
+        }
+
+        throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, "Wrong on, it should be MAXSIZE");
     }
 
-    private void consumeKey(ImapRequestLineReader request, GetAnnotationRequest getAnnotationRequest) throws DecodingException {
-        getAnnotationRequest.setKeys(ImmutableSet.of(request.atom()));
-        request.eol();
+    private void consumeKey(ImapRequestLineReader requestReader, GetAnnotationRequest.GetAnnotationRequestBuilder builder) throws DecodingException {
```

Comment:
This method is a side effect.

I would prefer return the key, and add to the builder in the calling code